text/0225-configuration.md

```diff
@@ -0,0 +1,411 @@
+
+## Explanation
+
+Using a configuration model or configuration file, users can configure all options currently available via environment variables.
+
+### Goals
+
+* The configuration must be language implementation agnostic. It must not contain structure or statements that only can be interpreted in a subset of languages. This does not preclude the possibility that the configuration can have specific extensions included for a subset of languages, but it does mean that the configuration must be interpretable by all implementation languages.
+* Broadly supported format. Ideally, the information encoded in the file can be decoded using native tools for all OpenTelemetry implementation languages. However, it must be possible for languages that do not natively support an encoding format to write their own parsers.
+* The configuration format must support structured data. At the minimum arrays and associative arrays.
+* The format must support at least null, boolean, string, double precision floating point (IEEE 754-1985), or signed 64 bit integer value types.
+* Custom span processors, exporters, samplers, or other user defined extension components can be configured using this format.
+* Configure SDK, but also configure instrumentation.
+* Must offer stability guarantees while supporting evolution
+* The structure of the configuration can be validated via a schema.
+* Support environment variable substitution to give users the option to avoid storing secrets in these files.
+
+## Internal details
+
+The schema for OpenTelemetry configuration is to be published in a repository to allow language implementations to leverage that definition to automatically generate code and/or validate end-user configuration. This will ensure that all implementations provide a consistent experience for any version of the schema they support. An example of such a proposed schema is available [here](https://github.com/MrAlias/otel-schema/tree/main/json_schema/schema).
+
+The working group proposes the use of [JSON Schema](https://json-schema.org/) as the language to define the schema. It provides:
+
+* support for client-side validation
+* code generation
+* broad support across languages
+
+In order to provide a minimal API surface area, implementations *MUST* support the following methods.
+
+### Configure(config)
+
+An API called `Configure` receives a configuration object. This method then applies the configuration object's details to the SDK. This method specifically applies the configuration object to allow for multiple configuration format providers to be supported in the future. This OTEP describes two such providers in a file and data structure formats below, but remote file formats *MAY* be implemented in the future.
```

Comment:
Let's provide an end-to-end example.

Reply:
> at the moment most SDKs provide imperative constructors for various SDK objects, and sometimes those constructors do a look-aside for env vars.

I think the lack of specificity on this front is a mistake, and I'd like to see file based configuration be more opinionated. 

All languages have imperative constructors. I'd like to see this "configure" API invoke those imperative constructors and return the configured tracer / meter / logger providers. In other words, it's not the responsibility of the providers to be aware of file based config. Separate code validates / parses the files and invokes the constructor APIs as necessary to reflect the configuration.

Reply:
So there should be a Configuration component which sets up the SDK components or an SDK component which encompasses "the SDK" and has the specified configuration object?

Reply:
> So there should be a Configuration component which sets up the SDK components 

This. Small clarification I don't have any opinion on how the code that does this is packaged / published (just in case there's any implication in the word "component").

Reply:
I specifically used the word "component" because it was the most general I could think of

Reply:
I've added some new language to try to improve clarity [here](https://github.com/open-telemetry/oteps/pull/225/commits/d1cf98cb222e3c367d999efed18c4590b630b9d3). 

Marking this issue as resolved, but please reopen if you feel its not adequately addressed.